### PR TITLE
[WIP] Initial cython implementation and proposed reorganization

### PR DESCRIPTION
**Pull request by dantegd**
This PR inlcudes the initial cython binding implementation for migration from CFFI. This PR is for discussion of both features, so it includes the  current `libgdf` code for testing while https://github.com/rapidsai/cudf/pull/318 is being worked at. But once the discussion is done this PR can either be rebased on 318 or the work of this PR can be pushed to 318 to avoid multiple merges. The reason for a separate PR is to avoid mixing discussions. 

Note that the library structure is independent to the bindings code, so we can change it without breaking things. 

Considerations of the new cython bindings:

1. The way they are structured, allows for a continuous migration without breaking any existing cffi functionality. The cython files (and therefore connection to `cpp` are contained in the `cudf/bind` folder that gets build to an extension. The file `cudf_cpp.pxd` is the header file that connects to the headers in `cudf-cpp` and contains useful functions for the bindings, such as translating a cudf-python column into a `gdf_column` struct for cpp and error handling. 

The 3 methods in the initial commit: `join` `reduce` and `sort` have been at well tested and show no performance regression in initial tests. The names of the files try to mirror `cudf-cpp` functionality that is being exposed (essentially decomposing the prior `_gdf.py`). 

Further methods already implemented that will be commited: `group_by`, `hash`, `nvtx`, `ops`, `quantile`. This covers a majority of the current cffi functionality excluding the csv reader and RMM. 

2. The bindings are independent enough of the `cpp` side that compilation times of the python library increased minimally when changing binding code, and not at all when doing pure python changes. 

3. To fully compile the bindings there is only one extra step: a. compile libgdf as it was before. b. To build the bindings then run `python setup.py build_ext --inplace`. c. Then the normal `python setup.py install`. 

The python part requires some minimal changes to some `setup.py` files to work with the new cudf-cpp, which will be changed once 318 progresses as well and this PR is rebased or incorporated into that one. 

Considerations for new structure:

To create a balance between understability of the cudf-python library, trying to be coherent with the new structure being proposed by PR 318 by propose a 1 level folder hierarchy that also doesn't deviate too much from pure flatness. The folders are:

- `bind` containing all cython code wrapping `cudf-cpp`
- `comm` containing all scripts regarding communications (like IPC related files)
- `dataframe` containing the data structures that are the core components of the cudf DataFrame and Series. 
- `groupby` feature folder containing the scripts for current groupby and older groupby scripts. Features that have significant code beyond bindings and code in `dataframe` scripts get their own folder.
- `io` All io related files (currently only `csv`)
- `tests` as it was
- `utils` All kind of utils functions such as `cudautils.py` and `queryutils.py`. 
- Some essential files in the root directory such as `settings.py` or `version.py`

This structure is based on the structure being proposed in PR 318 and also takes inspiration of the cupy and dask structures in some ways. Some more code modifications would be good (like renaming `columnops` to something like `typed_column`, but that can come in future PRs.

### CODE REVIEW

**Comment by dantegd** [comment]
Note: This PR will fail any kinds of CI since it adds the cython requirement and (currently) an additional step in `setup.py `

File: cudf/bind/cudf_cpp.pxd [hunk 1]
```diff
@@ -0,0 +1,1004 @@
...
+cdef gdf_context* create_context_view(flag_sorted, method, flag_distinct,
+                                      flag_sort_result, flag_sort_inplace)
+
+cpdef check_gdf_error(errcode)
+
+# Import cudf.h header to import all functions
+# First version of bindings has no changes to the cudf.h header, so this file
+# mirrors the structure in cpp/include
+
+cdef extern from "gdf.h" nogil:
```
Comment (harrism) [inline cudf/bind/cudf_cpp.pxd]: Is it possible to generate this from the libgdf headers like we did with cffi?

File: cudf/bind/cudf_cpp.pxd [hunk 2]
```diff
@@ -0,0 +1,1004 @@
...
+    cdef gdf_error gdf_segmented_radixsort_generic(gdf_segmented_radixsort_plan_type *hdl,
+                                         gdf_column *keycol, gdf_column *valcol,
+                                         unsigned num_segments,
+                                         unsigned *d_begin_offsets,
+                                         unsigned *d_end_offsets)
+
+    # // joins
+
+
+    # /* --------------------------------------------------------------------------*/
```
Comment (harrism) [inline cudf/bind/cudf_cpp.pxd]: Why duplicate the block comments here if they are already in the C/C++ headers? Just makes for more duplication to maintain.

File: cudf/bind/cudf_cpp.pxd (hunk 1 above)
Comment (dantegd) [reply cudf/bind/cudf_cpp.pxd]: yes, it is, a generating script is in the works for that. There are a few kinks we're still working on before adding that script.

File: cudf/bind/cudf_cpp.pxd (hunk 2 above)
Comment (dantegd) [reply cudf/bind/cudf_cpp.pxd]: I left them since during development it was useful to navigate the header file quickly, will delete them for the final PR.

**Pull request closed without merging**